### PR TITLE
Added basic harness for TeraChem Protocol Buffer Server mode.

### PR DESCRIPTION
**Pull request by coltonbh**
## Description
Added a basic `ProgramHarness`  for TeraChem's "server mode".

## Changelog description
Added a basic `ProgramHarness`  for TeraChem's "server mode".

## Status

- [ x] Code base linted
- [ ] Ready to go

Given this is my first PR please let me know what else your team would like to see! I'm trying to get an MVP version up and working so the code you see is rather minimal. The underlying library for communicating with the server--`tcpb`--needs a little love. I'll be patching it up to have better error handling over the coming weeks and intend to expand this `ProgramHarness` to reflect these changes once made.

Let me know what else your team would need! I'm happy to leave this PR open until a more robust harness is built (a couple of weeks, most likely), but I wanted to get eyes on as soon as possible. Thanks!

### CODE REVIEW

**Comment by coltonbh** [comment]
Closes #289 

Thanks, @loriab for your comments on making this PR. Please let me know what else is best practice for `QCEngine`. Thanks!

File: examples/terachem_pbs.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+"""
+)
+
+inp = qcel.models.AtomicInput(
+    molecule=mol,
+    driver="energy",
+    model={"method": "pbe0", "basis": "6-31g"},
+)
+ret = prog.compute(inp)
+print(ret)
```
Comment (loriab) [inline examples/terachem_pbs.py:25]: I didn't remember that there was an examples/ directory. :-)

Go ahead and add terachem pbs to https://github.com/MolSSI/QCEngine/blob/master/qcengine/testing.py#L142 and add some test cases to qcengine/programs/tests/ with the `@using("terachempbs")` guard. At a minimum, it'd be  good to add terachempbs to https://github.com/MolSSI/QCEngine/blob/master/qcengine/tests/test_harness_canonical.py#L13 and https://github.com/MolSSI/QCEngine/blob/master/qcengine/tests/test_harness_canonical.py#L82 lists. This harness can't be covered in CI, but at least you and perhaps others will notice problems locally in pytest.

File: qcengine/programs/terachem_pbs.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                raise ValueError(msg) from e
+        except ServerError as e:
+            msg = (
+                f"Unable to connect to TeraChem server at "
+                f"{os.getenv('TERACHEM_PBS_HOST')}:{os.getenv('TERACHEM_PBS_PORT')}"
+            )
+            logger.error(msg)
+            if raise_error:
+                raise OSError(msg) from e
+        return False
```
[the remark this answers is not in the document]
Comment (coltonbh) [reply qcengine/programs/terachem_pbs.py:80]: Great! I'll take a look at this.

File: examples/terachem_pbs.py (hunk 1 above)
Comment (coltonbh) [reply examples/terachem_pbs.py:25]: Will do!

**Comment by coltonbh** [comment]
@loriab  thanks for the review. I'll get to this a bit later this week (have a busy week!). Good to know what basic features you want tested and functional.

File: qcengine/programs/terachem_pbs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (coltonbh) [reply qcengine/programs/terachem_pbs.py:80]: Ok this has been addressed! I've added a `.get_version()` method so that a version is correctly displayed. Since communication with Terachem in the "protocol buffer" mode requires the `tcpb` python package I am checking for its existence as part of the `.found()` method. The behavior of `qcengine info` is as desired.

File: examples/terachem_pbs.py (hunk 1 above)
Comment (coltonbh) [reply examples/terachem_pbs.py:25]: @loriab not quite sure what to do here... Since you are unlikely to have a version of TeraChem running in the protocol buffer mode on your test server, should I add these tests? It seems you'd be unable to run them! Also, since the `tcpb` package is required, that's another stopper for running your tests against the TeraChem protocol buffer mode, unless we want to include the `tcpb` package as a requirement for `qcengine`, which we decided we didn't. I agree some tests of some sort would be good! But I'm not familiar enough with the package to see where to include tests that don't require actual computation or the additional `tcpb` package. Let me know your thoughts--thanks :)

File: qcengine/programs/terachem_pbs.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        """Returns version of TeraChem Protocol Buffer Server"""
+        try:
+            import tcpb
+        except ModuleNotFoundError:
+            return "???"
+        else:
+            try:
+                return tcpb.__version__
+            except AttributeError:
+                return "???"
```
[the remark this answers is not in the document]
Comment (coltonbh) [reply qcengine/programs/terachem_pbs.py:92]: I returned "???" because before I had the `get_version` method I saw this was what `QCEngine` returned by default. So I was sticking to that convention. Should I return something else instead?

File: qcengine/tests/test_harness_canonical.py [hunk 4]
```diff
@@ -21,6 +19,7 @@
     ("psi4", {"method": "hf", "basis": "6-31G"}, {}),
     ("qchem", {"method": "hf", "basis": "6-31G"}, {}),
     ("rdkit", {"method": "UFF"}, {}),
+    ("terachem_pbs", {"method": "b3lyp", "basis": "6-31G"}, {}),
```
[the remark this answers is not in the document]
Comment (coltonbh) [reply qcengine/tests/test_harness_canonical.py:22]: Added!

**Comment by coltonbh** [comment]
Added (simple) hf test!

**Comment by coltonbh** [comment]
Hi @loriab where do we stand on this? I think I've made all changes requested :) Just checking in so this doesn't get back-burnered indefinitely. Thanks for helping to shepherd this along!

File: qcengine/tests/test_harness_canonical.py [hunk 5]
```diff
@@ -127,6 +125,7 @@ def test_compute_energy_qcsk_basis(program, model, keywords):
         ("psi4", {"method": "bad"}),
         ("qchem", {"method": "bad"}),
         ("rdkit", {"method": "bad"}),
+        ("gcp", {"method": "bad"}),
```
Comment (loriab) [inline qcengine/tests/test_harness_canonical.py:128]:
```suggestion
        ("terachem_pbs", {"method": "bad"}),
```

File: qcengine/programs/terachem_pbs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (loriab) [reply qcengine/programs/terachem_pbs.py:92]: I'd feel better without those "???" around. Sorry that that poor example was there to be found. I think you can do like https://github.com/MolSSI/QCEngine/blob/master/qcengine/programs/dftd4.py#L47-L57 replacing `dftd4` by `tcpb`

**Review by loriab — COMMENTED** [review]
A couple items, then could you run all your tests locally again, please, since GH CI can't do them?

**Comment by coltonbh** [comment]
@loriab fixed! I believe I was following the "???" specification used in https://github.com/MolSSI/QCEngine/blob/7ca5ed0011f356d05a0951370fb9acb407f4558b/qcengine/cli.py#L89 and https://github.com/MolSSI/QCEngine/blob/7ca5ed0011f356d05a0951370fb9acb407f4558b/qcengine/cli.py#L109. Looks like just returning "None" is the better option and I can let the cli fill in the "???" when None is returned.

I had tried the higher level methods that you suggest from the `dftd4.py`'s `get_version` method. However, they failed a few edge cases since it's possible `tcpb` exists but the version is too old to have the attribute etc... Can't remember exactly what all the edge cases were 😬 but code as written correctly produced the correct behavior under all cases. I was finding it quite hard to mock out modules with/without attributes for tests so I just punted on putting those in tests since most other harnesses seemed to lack tests for their `get_version` method and the juice didn't seem worth the squeeze.

That said, the code as written works under all cases and I recall running into issues with the `which_program` and `safe_version` higher-level methods. I'd prefer to leave it (with the "???" removed), but can go over it again with the higher level methods to see what breaks again if you'd like me to dig deeper on this.

Let me know! Or if all looks good you can take it from here :)

Tests pass locally!

**Review by loriab — APPROVED** [review]
Thanks for the clarification on version checking and for the nice new harness.

GHA is failing at the moment, but that's no fault of the PR.

**Comment by coltonbh** [comment]
WOOT WOOT!!! 🥳

**Comment by coltonbh** [comment]
closes #284
